citest/gcp_testing/resource_snapshot.py

```diff
@@ -0,0 +1,794 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Collects resources managed by Google APIs.
```

Comment:
Because this finds everything and anything. It wont just tell you that "something" is used, but will tell you exactly what. And knowing what it is provides the ability to also delete it if you want. And since all that is in there, you could also take a general inventory and clean things up as you see them from one place rather than hunting around within pantheon (the latter isnt implemented but all the bits are there to put a new UI around it).

Also, since it works with anything, it isnt limited to what quotas capture. For example files in a bucket or BigQuery tables  or anything else. One of the problems it has is that it picks up operations you performed. It turns out you can delete these, though that results in an Operation of its own.

You wouldnt want to blindly trust the diff to delete because there could be other concurrent activity creating things. But once you review the diff, it would normally be easy. So at the beginning and end of the test we capture a snapshot of [pre/post]_$BUILD_NUMBER. Most likely nothing else is happening. Then after the build fails and we finish investigating, you can just compare and delete that diff independent of what happened since. If I just captured quotas, we'd know it leaked, but of course it leaked since we left around the artifacts to investigate.